File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    def dfs(idx, cnt):
        nonlocal answer
        if cnt == target and idx == len(numbers)-1:
            answer += 1
            return
        if idx+1 >= len(numbers):
            return
        dfs(idx+1, cnt+numbers[idx+1])
        dfs(idx+1, cnt-numbers[idx+1])
        
    dfs(0, numbers[0])
    dfs(0, numbers[0]*(-1))
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int N = 0;
    static int answer = 0;
    public int solution(int[] numbers, int target) {
        N = numbers.length;
        
        dfs(0, -1*numbers[0], target, numbers);
        dfs(0, numbers[0], target, numbers);
        
        return answer;
    }
    
    public void dfs(int idx, int result, int target, int[] numbers) {
        if (idx == N-1 && result == target) {
            answer++;
            return;
        }
        if (idx >= N-1) {
            return;
        }
        if (idx + 1 < N) {
            dfs(idx+1, result+numbers[idx+1], target, numbers);
            dfs(idx+1, result-numbers[idx+1], target, numbers);
        }
    }
}